thetis/callback.py

```diff
@@ -529,6 +529,20 @@ def __init__(self, solver_obj,
         self.field_names = field_names
         self._name = name
 
+        # create VertexOnlyMesh and function spaces for interpolation
+        self.mesh = solver_obj.fields[field_names[0]].function_space().mesh()
+        self.vom = VertexOnlyMesh(self.mesh, detector_locations, redundant=True)
+        self.function_spaces = {}
+        for field_name in field_names:
+            field = solver_obj.fields[field_name]
+            if isinstance(field.function_space().ufl_element(), VectorElement):
+                P0DG = VectorFunctionSpace(self.vom, "DG", 0)
+                P0DG_input_ordering = VectorFunctionSpace(self.vom.input_ordering, "DG", 0)
+            else:
+                P0DG = FunctionSpace(self.vom, "DG", 0)
+                P0DG_input_ordering = FunctionSpace(self.vom.input_ordering, "DG", 0)
+            self.function_spaces[field_name] = (P0DG, P0DG_input_ordering)
```

Comment:
minor: there's quite a lot of duplicated code in the vom creation and evaluation. Could plausibly be refactored in to helper functions for example

Reply:
I've created a new function in utility.py [L1159](https://github.com/thetisproject/thetis/pull/373/files#diff-a2f80a5edab6cf135404e407d32a7cce2cbbd4e911b63dc1b13dce5e65e50ee1R1159): `vom_interpolator_function`. It generates and returns the Functions only as suggested so we remove all the unnecessary Function duplication I had. We could probably go a step further and create a vom_interpolator class which also performs the interpolation as well rather than repeating these lines:

```
f_at_points, f_at_input_points = self.interp_functions[field_name]
f_at_points.interpolate(field)
f_at_input_points.interpolate(f_at_points)
return f_at_input_points.dat.data_ro[:]
```

Though maybe it's fine how it is for now. When [Firedrake issue 3080](https://github.com/firedrakeproject/firedrake/issues/3080) is dealt with we can probably revert to `.at()` anyway.

Reply:
Thanks, this is good for now